applications/vanilla/views/discussion/helper_functions.php

```diff
@@ -85,11 +85,17 @@ function writeComment($comment, $sender, $session, $currentOffset) {
         $sender->EventArguments['Permalink'] = $permalink;
 
         // Needed in writeCommentOptions()
-        if ($sender->data('Discussion', null) === null) {
+        $discussion = $sender->data('Discussion', null);
+        
+        if ($discussion === null) {
             $discussionModel = new DiscussionModel();
             $discussion = $discussionModel->getID($comment->DiscussionID);
             $sender->setData('Discussion', $discussion);
         }
+        
+        if (val('InsertUserID', $discussion) === $comment->InsertUserID) {
```

Comment:
I know we haven't messaged it publicly right now, but at the moment we are trying to deprecate and remove many usages of the val function and put the onus on code to know what it's dealing with. Something like the following would align with that better:

```php
$insertUserID = is_array($discussion) ? $discussion['InsertUserID'] : $discussion->InsertUserID;
if ($insertUserID === $comment->InsertUserID) {

```

Reply:
How about:

    $originalPosterID = $discussion->InsertUserID ?? $discussion['InsertUserID'] ?? false;

While there shouldn't ever be a comment without a discussion, this is just a render function that should work independently of outside data existing or not.

Reply:
This function is certainly poorly architected.

As for you suggested solution:

```php
$thing = new \stdClass();
$thing->test = "test";
$other = $thing['test'] ?? $things->test ?? false;
// Cannot use object of type stdClass as array

$thing = ['test' => 'test'];
$other = $things->test ?? $thing['test'] ?? false;
// Works??
```

It seems to work the second way but not the first, so I guess that is a bit cleaner than the `is_array` or `is_object` check. Please be sure to accompany it with a comment of why we need this check in the first place.

Reply:
It does work, but I couldn't find anything in the PHP documentation about it.
Is is evaluated in this order `a ?? (b ?? c)`. So clearly there is the possibility of object access on an array, which works, but there is no guarantee this will keep working in the future.

So i settled on just using Controller::data() (which uses valr internally, but avoids array/object confusion here, once it has been refactored).